src/blocks/calendar.rs

```diff
@@ -0,0 +1,385 @@
+//! Calendar
+//!
+//! This block displays upcoming calendar events retrieved from a CalDav ICalendar server.
+//!
+//! # Configuration
+//!
+//! Key | Values | Default
+//! ----|--------|--------
+//! `next_event_format` | A string to customize the output of this block when there is a next event in the calendar. See below for available placeholders. | <code>\" $icon $start.datetime(f:'%a %H:%M') $summary \"</code>
+//! `ongoing_event_format` | A string to customize the output of this block when an event is ongoing. | <code>\" $icon $summary (ends at $end.datetime(f:'%H:%M')) \"</code>
+//! `no_events_format` | A string to customize the output of this block when there are no events | <code>\" $icon \"</code>
+//! `interval` | Update interval in seconds | `60`
+//! `url` | CalDav calendar server URL | N/A
+//! `auth` | Authentication configuration (unauthenticated, basic, or oauth2) | `unauthenticated`
+//! `calendars` | List of calendar names to monitor. If empty, all calendars will be fetched. | `[]`
+//! `events_within_hours` | Number of hours to look for events in the future | `24`
+//! `warning_threshold` | Warning threshold in seconds for the upcoming event | `300`
+//! `browser_cmd` | Command to open event details in a browser. The block passes the HTML link as an argument | `"xdg-open"`
+//!
+//! Action          | Description                               | Default button
+//! ----------------|-------------------------------------------|---------------
+//! `open_link` | Opens the HTML link of the event | Left
+//!
+//!//! # Examples
+//!
+//! ## Unauthenticated
+//!
+//! ```toml
+//! [[block]]
+//! block = "calendar"
+//! next_event_format = " $icon $start.datetime(f:'%a %H:%M') $summary "
+//! ongoing_event_format = " $icon $summary (ends at $end.datetime(f:'%H:%M')) "
+//! no_events_format = " $icon no events "
+//! interval = 30
+//! url = "https://caldav.example.com/calendar/"
+//! calendars = ["user/calendar"]
+//! events_within_hours = 48
+//! ignore_all_day_events = true
+//! warning_threshold = 600
+//! browser_cmd = "firefox"
+//! [block.auth]
+//! type = "unauthenticated"
+//! ```
+//!
+//! ## Basic Authentication
+//!
+//! ```toml
+//! [[block]]
+//! block = "calendar"
+//! next_event_format = " $icon $start.datetime(f:'%a %H:%M') $summary "
+//! ongoing_event_format = " $icon $summary (ends at $end.datetime(f:'%H:%M')) "
+//! no_events_format = " $icon no events "
+//! interval = 30
+//! url = "https://caldav.example.com/calendar/"
+//! calendars = [ "Holidays" ]
+//! events_within_hours = 48
+//! ignore_all_day_events = true
+//! warning_threshold = 600
+//! browser_cmd = "firefox"
+//! [block.auth]
+//! type = "basic"
+//! username = "your_username"
+//! password = "your_password"
+//! ```
+//! Note: The `username` and `password` can also be provided by setting the environment variables `I3RS_CALENDAR_AUTH_USERNAME` and `I3RS_CALENDAR_PASSWORD` respectively.
+//!
+//! ## OAuth2 Authentication (Google Calendar)
```

Comment:
Should add a description about how to create the app and that you have to enable to caldav api (https://console.cloud.google.com/apis/library/caldav.googleapis.com)